linux/libcmt/src/ioctl/rollup-driver.c

```diff
@@ -0,0 +1,101 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdint.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <rollup-driver.h>
+#include <linux/cartesi/rollup.h>
+
+int cmt_rollup_driver_init(struct cmt_rollup_driver *me)
+{
+	int rc = 0;
+
+	me->fd = open("/dev/rollup", O_RDWR);
+	if (me->fd < 0)
+		return -errno;
+
+	struct rollup_setup setup;
+	if (ioctl(me->fd, IOCTL_ROLLUP_SETUP, &setup)) {
+		rc = -errno;
+		goto setup_failed;
+	}
+
+	me->tx->length = setup.tx.length;
+	me->tx->data = mmap((void *)setup.tx.data, setup.tx.length,
+	                    PROT_READ | PROT_WRITE,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->tx->data == MAP_FAILED) {
+		rc = -errno;
+		goto tx_map_failed;
+	}
+
+	me->rx->length = setup.rx.length;
+	me->rx->data = mmap((void *)setup.rx.data, setup.rx.length,
+	                    PROT_READ,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->rx->data == MAP_FAILED) {
+		rc = -errno;
+		goto rx_map_failed;
+	}
+	return 0;
+
+rx_map_failed:
+	munmap(me->tx->data, me->tx->length);
+tx_map_failed:
+setup_failed:
+	if (close(me->fd))
+		return -errno;
+	return rc;
+}
+
+int cmt_rollup_driver_fini(struct cmt_rollup_driver *me)
+{
```

Comment:
I think all `_fini` should invalidate all internal structure fields. I know using `me` after finalizing it is stated as undefined, but we should try to minimize the amount of undefined things that could happen. Leaving this way it could crash or not, cleaning it up it's likely that any subsequent function call will either fail or crash.